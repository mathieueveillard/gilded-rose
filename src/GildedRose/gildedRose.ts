import { Item } from "../Items/item";
import { ItemType, Quality, SellIn } from "../Enums/enum";

export class GildedRose {
  items: Array<Item>;

  constructor(items: Array<Item> = []) {
    this.items = items;
  }

  updateQuality() {
    for (const item of this.items) {
      console.log(`Before updateQuality: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);

      switch (item.name) {
        case ItemType.AgedBrie:
          this.updateAgedBrie(item);
          break;
        case ItemType.BackstagePasses:
          this.updateBackstagePasses(item);
          break;
        case ItemType.Sulfuras:
          break;
        default:
          this.updateStandardItem(item);
          break;
      }

      console.log(`After updateQuality: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
    }

    return this.items;
  }

  private updateAgedBrie(item: Item) {
    console.log(`Before updateAgedBrie: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);

    item.sellIn -= 1;
    item.quality = Math.min(Quality.MAX, item.quality + Quality.INCREASE);

    console.log(`After updateAgedBrie: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
  }

  private updateBackstagePasses(item: Item) {
    console.log(`Before updateBackstagePasses: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);

    item.sellIn -= 1;

    if (item.sellIn < 0) {
      item.quality = Quality.MIN;
    } else if (item.sellIn <= SellIn.WARNING) {
      item.quality = Math.min(Quality.MAX, item.quality + Quality.INCREASE_DOUBLE);
    } else if (item.sellIn <= SellIn.SOON) {
      item.quality = Math.min(Quality.MAX, item.quality + Quality.INCREASE);
    }

    console.log(`After updateBackstagePasses: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
  }

  private updateStandardItem(item: Item) {
    console.log(`Before updateStandardItem: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
    
    item.sellIn -= 1;
    
    // Diminue la qualité deux fois plus rapidement si la date de vente est passée
    if (item.sellIn < 0) {
        item.quality = Math.max(Quality.MIN, item.quality - Quality.DECREASE * 2);
    } else {
        item.quality = Math.max(Quality.MIN, item.quality - Quality.DECREASE);
    }
    
    // Assurez-vous que la qualité ne dépasse pas 50
    if (item.quality > Quality.MAX) {
        item.quality = Quality.MAX;
    }

    console.log(`After updateStandardItem: name=${item.name}, sellIn=${item.sellIn}, quality=${item.quality}`);
}
}
