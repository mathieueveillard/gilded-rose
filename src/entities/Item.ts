class Item {
  private _name: string;
  private _sellIn: number;
  private _quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this._name = name;
    this._sellIn = sellIn;
    this._quality = quality;
  }

  get name(): string {
    return this._name;
  }

  set name(value: string) {
    this._name = value;
  }

  get sellIn(): number {
    return this._sellIn;
  }

  set sellIn(value: number) {
    this._sellIn = value;
  }

  get quality(): number {
    return this._quality;
  }

  set quality(value: number) {
    this._quality = value;
  }
}

export default Item;
