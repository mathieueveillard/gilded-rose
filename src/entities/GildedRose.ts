import Item from './Item';

class GildedRose {
  private readonly _items: Array<Item>;

  constructor(items: Array<Item> = []) {
    this._items = items;
  }

  get items(): Array<Item> {
    return this._items;
  }

  public updateQuality(): Array<Item> {
    this._items.forEach(item => {
      console.log(`Processing item: ${item.name} with sellIn: ${item.sellIn} and quality: ${item.quality}`);

      // Reduce sellIn first for all items except Sulfuras
      if (item.name !== "Sulfuras, Hand of Ragnaros") {
        console.log('Item is not Sulfuras, reducing sellIn by 1');
        item.sellIn--;
      }

      switch (item.name) {
        case "Aged Brie":
          console.log('Item is Aged Brie');
          this.updateAgedBrie(item);
          break;
        case "Backstage passes to a TAFKAL80ETC concert":
          console.log('Item is Backstage passes');
          this.updateBackstagePasses(item);
          break;
        case "Sulfuras, Hand of Ragnaros":
          console.log('Item is Sulfuras, no change required');
          break;
        case "Conjured":
          console.log('Item is Conjured');
          this.updateConjuredItem(item);
          break;
        default:
          console.log('Item is a regular item');
          this.updateNormalItem(item);
          break;
      }

      // Ensure quality remains within valid bounds
      if (item.quality < 0) item.quality = 0;
      if (item.quality > 50 && item.name !== "Sulfuras, Hand of Ragnaros") item.quality = 50;
    });

    return this._items;
  }

  private updateAgedBrie(item: Item) {
    if (item.quality < 50) {
      console.log('Increasing quality by 1 for Aged Brie');
      item.quality++;
    }
    if (item.sellIn < 0 && item.quality < 50) {
      console.log('SellIn is negative, increasing quality by 1 again for Aged Brie');
      item.quality++;
    }
  }

  private updateBackstagePasses(item: Item) {
    if (item.quality < 50) {
      console.log('Increasing quality by 1 for Backstage passes');
      item.quality++;
      if (item.sellIn < 11 && item.quality < 50) {
        console.log('SellIn is less than 11, increasing quality by 1 again');
        item.quality++;
      }
      if (item.sellIn < 6 && item.quality < 50) {
        console.log('SellIn is less than 6, increasing quality by 1 once more');
        item.quality++;
      }
    }
    if (item.sellIn < 0) {
      console.log('Setting Backstage passes quality to 0 after concert');
      item.quality = 0;
    }
  }

  private updateNormalItem(item: Item) {
    if (item.quality > 0) {
      console.log('Reducing quality by 1 for regular item');
      item.quality--;
    }
    if (item.sellIn < 0 && item.quality > 0) {
      console.log('SellIn is negative, reducing quality by 1 again for regular item');
      item.quality--;
    }
  }

  private updateConjuredItem(item: Item) {
    if (item.quality > 0) {
      console.log('Reducing quality by 2 for Conjured item');
      item.quality -= 2;
    }
    if (item.sellIn < 0 && item.quality > 0) {
      console.log('SellIn is negative, reducing quality by 2 again for Conjured item');
      item.quality -= 2;
    }
    if (item.quality < 0) {
      console.log('Quality is negative, setting it to 0');
      item.quality = 0;  // Ensure quality is never negative
    }
  }
}

export default GildedRose;
