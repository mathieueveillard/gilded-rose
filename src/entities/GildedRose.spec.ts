import runGoldenMaster from 'jest-golden-master';

import GildedRose from './GildedRose';
import Item from './Item';

describe('GildedRose', () => {

  it('should decrease SellIn and Quality for a normal item by 1', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Normal Item', 10, 20)]);
      store.updateQuality();
      expect(store.items[0].sellIn).toBe(9);
      expect(store.items[0].quality).toBe(19);
    });
  });

  it('should degrade quality twice as fast once the sell by date has passed', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Normal Item', 0, 20)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(18);
    });
  });

  it('should increase quality of Aged Brie as it gets older', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Aged Brie', 10, 20)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(21);
    });
  });

  it('should not change quality or sellIn of Sulfuras', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Sulfuras, Hand of Ragnaros', 10, 80)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(80);
      expect(store.items[0].sellIn).toBe(10);
    });
  });

  it('should increase quality of Backstage passes as SellIn approaches', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(21);
    });
  });

  it('should increase quality of Backstage passes by 2 when 10 days or less for SellIn', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 10, 20)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(22);
    });
  });

  it('should increase quality of Backstage passes by 3 when 5 days or less for SellIn', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 5, 20)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(23);
    });
  });

  it('should drop quality of Backstage passes to 0 after concert', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 0, 20)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(0);
    });
  });

  it('should degrade quality of Conjured items twice as fast', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Conjured', 10, 20)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(18);
    });
  });

  it('should not make quality negative', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Normal Item', 10, 0)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(0);
    });
  });

  it('should not increase quality of any item above 50', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Aged Brie', 10, 50)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(50);
    });
  });

  it('should degrade quality of Conjured items four times as fast after sell date', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Conjured', 0, 20)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(16);
    });
  });

  it('should reset quality to 50 if it exceeds 50 for non-Sulfuras items', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Aged Brie', -1, 51)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(50);
    });
  });

  it('should initialize GildedRose with empty items array using default constructor', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose();
      expect(store.items.length).toBe(0);
    });
  });

  it('should set quality to 0 if an item\'s quality becomes negative', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Test Item', 10, -5)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(0);
    });
  });

  it('should increase quality of Aged Brie by 2 when sellIn is negative and quality is less than 50', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Aged Brie', -1, 48)]);
      store.updateQuality();
      expect(store.items[0].quality).toBe(50);
    });
  });

  it('should reset quality to 0 if a Conjured item goes below 0', () => {
    runGoldenMaster(async () => {
      const store = new GildedRose([new Item('Conjured', 10, 1)]);
      store.updateQuality();  // After updating, quality should be -1, but the check should reset it to 0
      expect(store.items[0].quality).toBe(0);
    });
  });


});
