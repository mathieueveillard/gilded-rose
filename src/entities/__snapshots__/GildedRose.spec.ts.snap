// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GildedRose should decrease SellIn and Quality for a normal item by 1 1`] = `
[
  "Processing item: Normal Item with sellIn: 10 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is a regular item",
  "Reducing quality by 1 for regular item",
]
`;

exports[`GildedRose should degrade quality of Conjured items four times as fast after sell date 1`] = `
[
  "Processing item: Conjured with sellIn: 0 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Conjured",
  "Reducing quality by 2 for Conjured item",
  "SellIn is negative, reducing quality by 2 again for Conjured item",
]
`;

exports[`GildedRose should degrade quality of Conjured items twice as fast 1`] = `
[
  "Processing item: Conjured with sellIn: 10 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Conjured",
  "Reducing quality by 2 for Conjured item",
]
`;

exports[`GildedRose should degrade quality twice as fast once the sell by date has passed 1`] = `
[
  "Processing item: Normal Item with sellIn: 0 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is a regular item",
  "Reducing quality by 1 for regular item",
  "SellIn is negative, reducing quality by 1 again for regular item",
]
`;

exports[`GildedRose should drop quality of Backstage passes to 0 after concert 1`] = `
[
  "Processing item: Backstage passes to a TAFKAL80ETC concert with sellIn: 0 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Backstage passes",
  "Increasing quality by 1 for Backstage passes",
  "SellIn is less than 11, increasing quality by 1 again",
  "SellIn is less than 6, increasing quality by 1 once more",
  "Setting Backstage passes quality to 0 after concert",
]
`;

exports[`GildedRose should increase quality of Aged Brie as it gets older 1`] = `
[
  "Processing item: Aged Brie with sellIn: 10 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Aged Brie",
  "Increasing quality by 1 for Aged Brie",
]
`;

exports[`GildedRose should increase quality of Aged Brie by 2 when sellIn is negative and quality is less than 50 1`] = `
[
  "Processing item: Aged Brie with sellIn: -1 and quality: 48",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Aged Brie",
  "Increasing quality by 1 for Aged Brie",
  "SellIn is negative, increasing quality by 1 again for Aged Brie",
]
`;

exports[`GildedRose should increase quality of Backstage passes as SellIn approaches 1`] = `
[
  "Processing item: Backstage passes to a TAFKAL80ETC concert with sellIn: 15 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Backstage passes",
  "Increasing quality by 1 for Backstage passes",
]
`;

exports[`GildedRose should increase quality of Backstage passes by 2 when 10 days or less for SellIn 1`] = `
[
  "Processing item: Backstage passes to a TAFKAL80ETC concert with sellIn: 10 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Backstage passes",
  "Increasing quality by 1 for Backstage passes",
  "SellIn is less than 11, increasing quality by 1 again",
]
`;

exports[`GildedRose should increase quality of Backstage passes by 3 when 5 days or less for SellIn 1`] = `
[
  "Processing item: Backstage passes to a TAFKAL80ETC concert with sellIn: 5 and quality: 20",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Backstage passes",
  "Increasing quality by 1 for Backstage passes",
  "SellIn is less than 11, increasing quality by 1 again",
  "SellIn is less than 6, increasing quality by 1 once more",
]
`;

exports[`GildedRose should initialize GildedRose with empty items array using default constructor 1`] = `[]`;

exports[`GildedRose should not change quality or sellIn of Sulfuras 1`] = `
[
  "Processing item: Sulfuras, Hand of Ragnaros with sellIn: 10 and quality: 80",
  "Item is Sulfuras, no change required",
]
`;

exports[`GildedRose should not increase quality of any item above 50 1`] = `
[
  "Processing item: Aged Brie with sellIn: 10 and quality: 50",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Aged Brie",
]
`;

exports[`GildedRose should not make quality negative 1`] = `
[
  "Processing item: Normal Item with sellIn: 10 and quality: 0",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is a regular item",
]
`;

exports[`GildedRose should reset quality to 0 if a Conjured item goes below 0 1`] = `
[
  "Processing item: Conjured with sellIn: 10 and quality: 1",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Conjured",
  "Reducing quality by 2 for Conjured item",
  "Quality is negative, setting it to 0",
]
`;

exports[`GildedRose should reset quality to 50 if it exceeds 50 for non-Sulfuras items 1`] = `
[
  "Processing item: Aged Brie with sellIn: -1 and quality: 51",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is Aged Brie",
]
`;

exports[`GildedRose should set quality to 0 if an item's quality becomes negative 1`] = `
[
  "Processing item: Test Item with sellIn: 10 and quality: -5",
  "Item is not Sulfuras, reducing sellIn by 1",
  "Item is a regular item",
]
`;
